File: playgrounds/subclass_sandbox/subclass_sandbox.js
class Superpower {
  //   constructor(soundEngine) {
  //     this.soundEngine = soundEngine;
  //   }

  activate() {
    throw new Error("this has not been implented");
  }

  move(x, y, z) {
    // move the player
  }

  playSound(soundId, volume) {
    // play the sound
    this.soundEngine.play(soundId, volume);
  }

  spawnParticles(particleType, count) {
    // display the particles
  }

  getHeroX() {
    // get the x position of the hero
  }

  getHeroY() {
    // get the y position of the hero
  }

  getHeroZ() {
    // get the z position of the hero
  }

  getSoundPlayer() {
    return state.soundPlayer;
  }
}

class SkyLaunch extends Superpower {
  constructor() {
    super();
  }

  activate() {
    if (this.getHeroY() == 0) {
      // standing on the ground
      this.playSound("sproing", 1);
      this.spawnParticles("dust", 200);
      this.move(0, 20, 0);
    } else if (this.getHeroY() < 10) {
      // in the air, but not high up
      this.playSound("swoop", 1);
      this.spawnParticles("cloud", 50);
      this.move(0, this.getHeroY() - 10, 0);
    } else {
      // we are high in the air
      this.playSound("dive", 1);
      this.move(0, 0, 0);
    }
  }
}

const skylaunch = new SkyLaunch();
skylaunch.activate();

File: playgrounds/subclass_sandbox/subclass_sandbox.rs
trait SuperPower {
    fn activate(&mut self, hero: &mut Hero, sound_system: &mut SoundSystem);

    fn move_hero(&mut self, x: f32, y: f32, z: f32, hero: &mut Hero) {
        // move the hero
        hero.apply_force([x, y, z]);
    }

    fn play_sound(&self, soundId: i32, volume: f32, sound_system: &mut SoundSystem) {
        // play this sound by default
        sound_system.playsound(soundId, volume);
    }
}

struct SkyLaunch {}

impl SuperPower for SkyLaunch {
    fn activate(&mut self, hero: &mut Hero, sound_system: &mut SoundSystem) {
        // launch into the sky
        self.moveHero(0.0, 20.0, 0.0, &mut hero);
        self.play_sound(JUMP_SOUND, 1.0, &mut sound_system);
    }
}